vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+  return &ClusterState{
+     RecommenderName:   recommender_name,
+     Pods:              make(map[PodID]*PodState),
+     Vpas:              make(map[VpaID]*Vpa),
+     EmptyVPAs:         make(map[VpaID]time.Time),
+     aggregateStateMap: make(aggregateContainerStatesMap),
+     labelSetMap:       make(labelSetMap),
+  }
+}
+```
+
+Therefore, when loading VPA objects to the `clusterStateFeeder`, it can use the field selector to select VPA CRDs that 
+have `recommenderName` equal to the current clusterState’s `RecommenderName`.
+```golang
+// Fetch VPA objects and load them into the cluster state.
+func (feeder *clusterStateFeeder) LoadVPAs() {
+  // List VPA API objects.
+  allVpaCRDs, err := feeder.vpaLister.List(labels.Everything())
+  if err != nil {
+     klog.Errorf("Cannot list VPAs. Reason: %+v", err)
+     return
+  }
+
+  var vpaCRDs []*vpa_types.VerticalPodAutoscaler
+  for _, vpaCRD := range allVpaCRDs {
+     currentRecommenderName := feeder.clusterState.RecommenderName
+     if (vpaCRD.Spec.RecommenderName != currentRecommenderName) && (vpaCRD.Spec.RecommenderName != "") {
+        klog.V(6).Infof("Ignoring the vpaCRD as its name %v is not equal to the current recommender's name %v", vpaCRD.Spec.RecommenderName, currentRecommenderName)
+        continue
+     }
+     vpaCRDs = append(vpaCRDs, vpaCRD)
+
+  klog.V(3).Infof("Fetched %d VPAs.", len(vpaCRDs))
+  // Add or update existing VPAs in the model.
+  vpaKeys := make(map[model.VpaID]bool)
+
+  …
+
+  feeder.clusterState.ObservedVpas = vpaCRDs
+}
+```
+
+Accordingly, the VPA object yaml should include the `recommenderName` as the default `RecommenderName`.
+```yaml
+apiVersion: "autoscaling.k8s.io/v1"
+kind: VerticalPodAutoscaler
+metadata:
+ name: hamster-vpa
+Spec:
+ recommenderName: default
+ targetRef:
+   apiVersion: "apps/v1"
+   ... ... 
+```
+
+## Design Details
+
+### Test Plan
+- Add e2e test demonstrating the default recommender ignores a VPA which specifies an alternate recommender.
+
+### Upgrade / Downgrade Strategy
+For cluster upgrades, the VPA from the previous version will continue working as before. 
+There is no change in behavior or flags which have to be enabled or disabled.
+
+## Alternatives
+
+### Develop a plugin framework for customizable recommenders.
+Add a webhook system for customized recommendations. The default VPA recommender would 
+call any available recommendation webhooks, and if any of them make a recommendation, 
+the recommender would use that recommendation instead of making its own. If none make 
+a recommendation, it would make its recommendation as it currently does.
+
+### Develop a label selector approach.
+Add a label for the CRD object to denote the recommender’s name. When making 
+recommendations in the recommender, only the VpaCrds with the label 
+`recommender=default` will be loaded and updated by the existing recommender. 
+A label selector approach is too powerful and users can easily ignore those labels 
```

Comment:
Would it still be a problem if we said that VpaCrds w/o any label are still handled by the default recommender?